### PR TITLE
WIP! Cherow 2.0 Beta

### PR DESCRIPTION
**Pull request by KFlash**
This is the Cherow 2.0 Beta code that will land mid of january 2019.

Changes:

- Implemented `onToken` and `onComment` 
- Option to disable web compability (*Annex B*)
- EcmaVersion option
- Fixed all known bugs from 1.6.x
- Added latest Stage 3 support
- Improved performance

*...to be updated..*

**TODO**

- [ ] Optimize scanner 

- [ ] Finish parser code

### CODE REVIEW

**Comment by aladdin-add** [comment]
great job! 👏

I've made a simple analysis of the bundle (cherow.umd.min.js). seems `src/unicode.ts`, `/src/scanner/regexp.ts`, and `/src/state.ts` takes the large body of the bundle.

<details>

```bash
-----------------------------
Rollup File Analysis
-----------------------------
bundle size:    117.397 KB
original size:  150.431 KB
code reduction: 21.96 %
module count:   17
-----------------------------
file:            /src/unicode.ts
bundle space:    29.95 %
rendered size:   35.16 KB
original size:   31.764 KB
code reduction:  0 %
dependents:      3
  - /src/scanner/common.ts
  - /src/chars.ts
  - /src/scanner/regexp.ts
-----------------------------
file:            /src/scanner/regexp.ts
bundle space:    23.74 %
rendered size:   27.872 KB
original size:   32.082 KB
code reduction:  13.12 %
dependents:      1
  - /src/scanner/scan.ts
-----------------------------
file:            /src/state.ts
bundle space:    10.55 %
rendered size:   12.382 KB
original size:   16.547 KB
code reduction:  25.17 %
dependents:      1
  - /src/cherow.ts
-----------------------------
file:            /src/scanner/scan.ts
bundle space:    9.25 %
rendered size:   10.86 KB
original size:   12.59 KB
code reduction:  13.74 %
dependents:      1
  - /src/scanner/index.ts
-----------------------------
file:            /src/scanner/string.ts
bundle space:    4.54 %
rendered size:   5.324 KB
original size:   6.683 KB
code reduction:  20.34 %
dependents:      2
  - /src/scanner/template.ts
  - /src/scanner/scan.ts
-----------------------------
file:            /src/scanner/common.ts
bundle space:    4.13 %
rendered size:   4.847 KB
original size:   6.163 KB
code reduction:  21.35 %
dependents:      6
  - /src/scanner/comments.ts
  - /src/scanner/string.ts
  - /src/scanner/template.ts
  - /src/scanner/regexp.ts
  - /src/scanner/numeric.ts
  - /src/scanner/scan.ts
-----------------------------
file:            /src/scanner/numeric.ts
bundle space:    3.18 %
rendered size:   3.739 KB
original size:   5.162 KB
code reduction:  27.57 %
dependents:      1
  - /src/scanner/scan.ts
-----------------------------
file:            /src/token.ts
bundle space:    2.76 %
rendered size:   3.245 KB
original size:   10.675 KB
code reduction:  69.6 %
dependents:      2
  - /src/scanner/identifier.ts
  - /src/state.ts
-----------------------------
file:            /src/scanner/comments.ts
bundle space:    2.68 %
rendered size:   3.146 KB
original size:   4.097 KB
code reduction:  23.21 %
dependents:      2
  - /src/scanner/scan.ts
  - /src/scanner/index.ts
-----------------------------
file:            /src/scanner/template.ts
bundle space:    2.46 %
rendered size:   2.891 KB
original size:   3.615 KB
code reduction:  20.03 %
dependents:      1
  - /src/scanner/scan.ts
-----------------------------
file:            /src/errors.ts
bundle space:    2.16 %
rendered size:   2.532 KB
original size:   4.617 KB
code reduction:  45.16 %
dependents:      8
  - /src/scanner/common.ts
  - /src/scanner/comments.ts
  - /src/scanner/string.ts
  - /src/scanner/template.ts
  - /src/scanner/regexp.ts
  - /src/scanner/numeric.ts
  - /src/scanner/identifier.ts
  - /src/common.ts
-----------------------------
file:            /src/cherow.ts
bundle space:    1.96 %
rendered size:   2.302 KB
original size:   5.759 KB
code reduction:  60.03 %
dependents:      0
-----------------------------
file:            /src/common.ts
bundle space:    1.02 %
rendered size:   1.202 KB
original size:   3.676 KB
code reduction:  67.3 %
dependents:      2
  - /src/state.ts
  - /src/cherow.ts
-----------------------------
file:            /src/scanner/identifier.ts
bundle space:    1.02 %
rendered size:   1.201 KB
original size:   1.71 KB
code reduction:  29.77 %
dependents:      1
  - /src/scanner/scan.ts
-----------------------------
file:            /src/chars.ts
bundle space:    0.39 %
rendered size:   456 Bytes
original size:   4.232 KB
code reduction:  89.22 %
dependents:      3
  - /src/scanner/regexp.ts
  - /src/scanner/numeric.ts
  - /src/scanner/identifier.ts
-----------------------------
file:            /src/scope.ts
bundle space:    0.2 %
rendered size:   238 Bytes
original size:   985 Bytes
code reduction:  75.84 %
dependents:      1
  - /src/cherow.ts
-----------------------------
file:            /src/scanner/index.ts
bundle space:    0 %
rendered size:   0 Byte
original size:   74 Bytes
code reduction:  100 %
dependents:      3
  - /src/common.ts
  - /src/state.ts
  - /src/cherow.ts
-----------------------------
```

</details>

**Comment by KFlash** [comment]
I'm, aware of the error, but right now I'm facing a very strange issue that prevent me from fixing it. Everything pass online and on your computer, but on my computer I have over 200 failing tests. It just the same files. No idea what's going on. Maybe a cache issue?

cherow

**Comment by aladdin-add** [comment]
did you rebuild with `npm run bundle` and try again?

**Comment by KFlash** [comment]
I didn't. I just run the tests. Anyway I pushed a fix for the ASI issue you faced. Should work for you at least :)

**Comment by KFlash** [comment]
@aladdin-add Are you sure that all tests on beta branch pass for you? If so I will reformat my computer and re-install everything and see if that fixes things

**Comment by aladdin-add** [comment]
yes, I confirmed, and the ci passed.

**Comment by KFlash** [comment]
Ok. I will buy a new computer tomorrow. Shame because the one I'm using now is only 3 weeks old, but it's time to throw it in the trash.

**Comment by aladdin-add** [comment]
emm.. it might not be the computer's issue, you tried to remove the `rpt2_cache`?

**Comment by KFlash** [comment]
Where is it located?

**Comment by aladdin-add** [comment]
the root of repo, it was hidden since started with `.`

**Comment by aladdin-add** [comment]
@KFlash prettier is working as expected, it didn't change the ignored files -- seems you deleted some ignored files in this commit(https://github.com/cherow/cherow/commit/f659e8b7711542af28c2029304872231d85e19dd), and added a changed version in later commits(https://github.com/cherow/cherow/commit/1486f5000d8164f9ef3e0149d88575f7d1781ad8)

**Comment by KFlash** [comment]
@aladdin-add You gave me a conflicting file now on this PR. Plan to fix it?

**Comment by KFlash** [comment]
@cherow/cherow-team I  have to admit that I locale have Cherow 2.0 fully working already. I finished it last month :) I'm only re-using that code base and fix the negative things I know about from my own Cherow 2.0 code base before pushing here.

**Comment by aladdin-add** [comment]
you can use `// prettier-ignore` : it will exclude the next node in the abstract syntax tree from formatting.

**Comment by KFlash** [comment]
@cherow/cherow-team Current state on Cherow 2.0 Beta

```js
   Statements - Try (pass)
         try {} catch (fkleuver) { const fkleuver = 1; } :
     SyntaxError: Line 1, column 45: Identifier 'fkleuver' has already been declared
```

**Comment by KFlash** [comment]
@aladdin-add @fkleuver This program `const a = b; let a = c` will throw in Cherow 2.0 beta, but passed just fine in 1.6x. What you guys think have happened?

**Comment by fkleuver** [comment]
Scoping huehuehue

**Comment by KFlash** [comment]
@fkleuver @aladdin-add You  are both right, and it's a breaking change in 2.0. We need to document this somehow.  And my next push will include complete lexical scoping implementation. Also `try/catch`, `blockstatement` and `switch statement` now works with block scoping. However I ran into a strange issue with the scanner. Ideas about this?

```js
   Statements - Block (pass)
         let {[x]: y = z} = a;:
     TypeError: table[first] is not a function
```

**Comment by KFlash** [comment]
@aladdin-add If you want to do some documentation you may also document the fact that this scoping implementation is AnnexB sensitive :) If you turn off web compability the behavior will change, and Acorn haven't implemented most of this. Same with ESLint. And the Acorn implementation try to only  follow AnnexB, but fails in many cases.

**Comment by fkleuver** [comment]
You can just turn it off with a flag, right? So then it's only breaking if you have it turned on

**Comment by KFlash** [comment]
@fkleuver What breaking? AnnexB?  It can be turned off

Btw. any idea what cause that table lookup issue in the scanner?

**Comment by fkleuver** [comment]
What issue?

**Comment by KFlash** [comment]
@fkleuver 

```js
   Statements - Block (pass)
         let {[x]: y = z} = a;:
     TypeError: table[first] is not a function
```

**Comment by KFlash** [comment]
@fkleuver @aladdin-add I have 3 huge promises regarding what should land in v. 2.0.

- [ ] Full scoping and lexical scoping

- [ ] Exported bindings (*module code*)

- [ ] Performance

I'm now going to complete my promises one by one :)

**Comment by KFlash** [comment]
@cherow/cherow-team Possible to get a review on existing code so far before I continue? At least you got a taste on how Cherow 2.0 will be. And **yes** I finished it actually last month :)

For now I need to clean up things and wait for reviews before finishing expression stuff

File: src/scanner/scan.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ import { scanIdentifier, scanMaybeIdentifier } from './identifier';
 // Table for one char punctuator lookup
 const OneCharPunc = new Array(128).fill(0) as Token[];
 
-const table = new Array(0xffff).fill(scanMaybeIdentifier, 0, 0x80) as ((
+const table = new Array(0xffff).fill(scanMaybeIdentifier, 0x80) as ((
```
Comment (fkleuver) [inline src/scanner/scan.ts]: lol

Comment (KFlash) [reply src/scanner/scan.ts]: Why?

**Comment by KFlash** [comment]
@fkleuver Tests pass again now. Not for you?

**Comment by KFlash** [comment]
@cherow/cherow-team Guys! 2 of my promises have been kept!  I just implemented exported bindings (*module code*).  And I'm going to stop add more things for now until the code are 100%  stable.

But yeah!  I'm going to keep all my promises before I finally retire :)

**Comment by KFlash** [comment]
@fkleuver You understand now what Cherow 2.0 will become soon? We already passed Acorn and 
Esprima :) So what more can I give?  Well....... I have a few more secrets left that you will find out when we start to implement the rest of expression parsing.

@aladdin-add We just got a new breaking change! Can you document it?

**Comment by KFlash** [comment]
@cherow/cherow-team Any review  on current code?

@fkleuver I'm moving forward and starting to finishing expressions so we can move to master. Can you try to optimize / clean up current code case?

**Comment by KFlash** [comment]
@aladdin-add @fkleuver Things get complicated when we try to avoid too much use of member access. Because of that we are passing the "scope", but it turns  out to be complicated inside several expression function such as arrow arg list. No worries. I know how to fix it, but I'm not doing it right now :)

For now this fails as expected `a => { let a;}`, but this doesn't fail `(a, a) => {}` because the identifier hasn't been added to the current scope. And this one `([a, a]) => {}` or even `([a, b]) => { let a; }` doesn't fail because it doesn't have the scope created with arrows.

I'm aware of all this, and also why it doesn't fail *yet*. I will fix this later on, but the scope things works and have been implemented!

**Comment by KFlash** [comment]
@cherow/cherow-team My last surprise! I aim to log out from Github for the very last time end of this month :) Cherow 2.0 is almost there now. Last thing to do is the class implementation, and fixing non-failing tests!!

**Comment by KFlash** [comment]
@fkleuver @aladdin-add My last promise to keep was performance. The scoping thing should have killed performance vs. 1.6.x.  And please also note that this is beta and not optimized at all :)

Did i keep all my promises?

last_promise

**Comment by KFlash** [comment]
This is more accurate benchmark result. Current Cherow 2.0 Beta is way faster than 1.6.x

last_promise

**Comment by KFlash** [comment]
@fkleuver @aladdin-add @g-plane Ok, guys! Cherow 2.0 is here now! Only left to implement is 
class expr  / decl. I'm going to start on that now. After we can move to Master branch and fix the non-failing tests!

**Comment by KFlash** [comment]
@fkleuver @aladdin-add The reason that the new class code is huge compared to 1.6.x is because of this. I'm going to optimize it further, but at least you guys understand why I did it like this.

class

**Comment by KFlash** [comment]
Closing and moving to Master branch for finale finalizing of the code

**Pull request merged**